V6 tee update (#62)
iExecBlockchainComputing/documentation: for-developers/confidential-computing/create-your-first-sgx-app.md

{% endcode %}
{% endtab %}
{% endtabs %}

## Build the TEE docker image:

You will need to register a [free SCONE Account](https://scontain.com) to access the SCONE build tools and currated images from the [SCONE registry](https://gitlab.scontain.com/).

```sh
# once registered run `docker login` to connect the SCONE registry
docker login registry.scontain.com:5050
```

We will use the following script to wrap the sconification process, copy the `sconify.sh` script in the current directory:

{% code title="sconify.sh" %}
```bash
#!/bin/bash

# declare the app entrypoint
ENTRYPOINT="node /app/app.js"
# declare an image name
IMG_NAME=nodejs-hello-world

IMG_FROM=${IMG_NAME}:temp-non-tee
IMG_TO=${IMG_NAME}:tee-debug

# build the regular non-TEE image
docker build . -t ${IMG_FROM}

# pull the SCONE currated image corresponding to our base image
docker pull registry.scontain.com:5050/sconecuratedimages/node:14.4.0-alpine3.11

# run the sconifier to build the TEE image based on the non-TEE image
docker run -it --rm \
            -v /var/run/docker.sock:/var/run/docker.sock \
            registry.scontain.com:5050/scone-production/iexec-sconify-image:5.3.7 \
            sconify_iexec \

>             --name=hello-world \

Reviewer: Fixed value, should it be ${IMG_NAME} ?

Author: good question `--name` is not documented, @jeremyjams thought?

Author: @luisccastillo I'll change it :+1:

Once you are happy with the debug app, contact-us to go to production!

{% endhint %}

## Test your app on iExec

At this stage, your application is ready to be tested on iExec. The process is similar to testing any type of application on the platform, with these minor exceptions:

### Push the newly built TEE image to dockerhub

When you tag your Docker image, use `my-tee-hello-world` instead of `my-hello-world`.

### Deploy the TEE app on iExec:

TEE applications require some additional information to be filled in during deployment.

```sh
# prepare the TEE application template
iexec app init --tee
```

Edit `iexec.json` and fill in the standard keys and the `mrenclave` object:

```json
{
  ...
  "app": {
    "owner": "0xF048eF3d7E3B33A465E0599E641BB29421f7Df92", // your address
    "name": "tee-hello-world", // application name
    "type": "DOCKER",
    "multiaddr": "docker.io/username/my-tee-hello-world:1.0.0", // app image
    "checksum": "0x15bed530c76f1f3b05b2db8d44c417128b8934899bc85804a655a01b441bfa78", // image digest
    "mrenclave": {
      "provider": "SCONE", // TEE provider (keep default value)

>       "version": "v5", // protocol version (keep default value)

Reviewer 2: ```suggestion
      "version": "v5", // Scone version (keep default value)
```

      "entrypoint": "node /app/app.js", // your app image entrypoint

>       "heapSize": 1073741824, // heap size in bytes (1Gb)

Reviewer 2: ```suggestion
      "heapSize": 1073741824, // heap size in bytes (1GB)
```

>       "fingerprint": "eca3ace86f1e8a5c47123c8fd271319e9eb25356803d36666dc620f30365c0c1" // fspf fingerprint see below

Reviewer 2: ```suggestion
      "fingerprint": "eca3ace86f1e8a5c47123c8fd271319e9eb25356803d36666dc620f30365c0c1" // fingerprint of the enclave code (mrenclave), see how to retrieve it below
```

    }
  },
  ...
}
```

{% hint style="info" %}

> Run your TEE image with `SCONE_HASH=1` to get the fspf fingerprint:

Reviewer 2: ```suggestion
Run your TEE image with `SCONE_HASH=1` to get the enclave fingerprint (mrenclave):
```

mkdir my-tee-hello-world-app && cd my-tee-hello-world-app
mkdir src
touch Dockerfile
touch sconify.sh
```

In the folder `src/` create the file `app.js` then copy [this](../your-first-app#write-the-app-javascript-script-example) code inside.

> As we mentioned earlier, the advantage of using **SCONE** is the ability to make the application **Intel® SGX-enabled** without changing the source code. The only thing we are going to do is rebuilding the app using the Trusted-Execution-Environment currated image provided by **SCONE**.

Reviewer 2: ```suggestion
As we mentioned earlier, the advantage of using **SCONE** is the ability to make the application **Intel® SGX-enabled** without changing the source code. The only thing we are going to do is rebuilding the app using the Trusted-Execution-Environment tooling provided by **SCONE**.
```

{% hint style="info" %}

> SCONE provides currated base images for a bunch of languages, the currated image will be used to replace the base image and get rid of any non TEE compliant operations.

Reviewer 2: ```suggestion
SCONE provides TEE conversion tooling (Python, Java, ..) plus eventually TEE base images for other languages (NodeJs).
```

{% endhint %}

Copy the Dockerfile of the non-TEE app:

{% tabs %}
{% tab title="Javascript" %}
{% code title="Dockerfile" %}
```bash
# Starting from a base image supported by SCONE  
FROM node:14-alpine3.11

# install your dependencies
RUN mkdir /app && cd /app && npm install figlet@1.x

COPY ./src /app

ENTRYPOINT [ "node", "/app/app.js"]
```
{% endcode %}
{% endtab %}
{% endtabs %}

## Build the TEE docker image:

> You will need to register a [free SCONE Account](https://scontain.com) to access the SCONE build tools and currated images from the [SCONE registry](https://gitlab.scontain.com/).

Reviewer 2: ```suggestion
You will need to register a [free SCONE Account](https://scontain.com) to access the SCONE build tools and curated images from the [SCONE registry](https://gitlab.scontain.com/).
```

{% code title="sconify.sh" %}
```bash
#!/bin/bash

# declare the app entrypoint
ENTRYPOINT="node /app/app.js"
# declare an image name
IMG_NAME=nodejs-hello-world

IMG_FROM=${IMG_NAME}:temp-non-tee
IMG_TO=${IMG_NAME}:tee-debug

# build the regular non-TEE image
docker build . -t ${IMG_FROM}

# pull the SCONE currated image corresponding to our base image
docker pull registry.scontain.com:5050/sconecuratedimages/node:14.4.0-alpine3.11

# run the sconifier to build the TEE image based on the non-TEE image
docker run -it --rm \
            -v /var/run/docker.sock:/var/run/docker.sock \
            registry.scontain.com:5050/scone-production/iexec-sconify-image:5.3.7 \
            sconify_iexec \
            --name=${IMG_NAME} \
            --from=${IMG_FROM} \
            --to=${IMG_TO} \
            --binary-fs \
            --fs-dir=/app \
            --host-path=/etc/hosts \
            --host-path=/etc/resolv.conf \
            --binary=/usr/local/bin/node \
            --heap=1G \
            --dlopen=2 \
            --no-color \
            --verbose \
            --command=${ENTRYPOINT} \

>             && echo && echo "successfully built TEE docker image => ${IMG_TO}"

Reviewer 2: Maybe print SCONE_HASH here?

Author: yes, added: f44792fe5755e07134a6034746833b845971fc36

    "checksum": "0x15bed530c76f1f3b05b2db8d44c417128b8934899bc85804a655a01b441bfa78", // image digest
    "mrenclave": {
      "provider": "SCONE", // TEE provider (keep default value)
      "version": "v5", // Scone version (keep default value)
      "entrypoint": "node /app/app.js", // your app image entrypoint
      "heapSize": 1073741824, // heap size in bytes (1GB)
      "fingerprint": "eca3ace86f1e8a5c47123c8fd271319e9eb25356803d36666dc620f30365c0c1" // fingerprint of the enclave code (mrenclave), see how to retrieve it below
    }
  },
  ...
}
```

{% hint style="info" %}
Run your TEE image with `SCONE_HASH=1` to get the enclave fingerprint (mrenclave):
```sh
docker run -it --rm -e SCONE_HASH=1 tee-nodejs-hello-world:tee-debug
```
{% endhint %}

Deploy the app with the standard command:
```sh
iexec app deploy --chain viviani
```

### Run the TEE app

Specify the tag `--tag tee` in `iexec app run` command to run a tee app.

One last thing, in order to run a **TEE-debug** app you will also need to select a debug workerpool, use the Viviani debug workerpool `0xe6806E69BA8650AF23264702ddD43C4DCe35CcCe` (see deployed workerpools on https://v6.pools.iex.ec).

> The debug workerpool is connected to a debug Secret Management System (this is fine for debuging but do not use to store production secrets), we will need to init the storage token on this SMS.

Reviewer 2: ```suggestion
The debug workerpool is connected to a debug Secret Management Service (this is fine for debugging but do not use to store production secrets), we will need to init the storage token on this SMS.
```

As we mentioned earlier, the advantage of using **SCONE** is the ability to make the application **Intel® SGX-enabled** without changing the source code. The only thing we are going to do is rebuilding the app using the Trusted-Execution-Environment tooling provided by **SCONE**.

{% hint style="info" %}

> SCONE provides TEE conversion tooling (Python, Java, ..) plus eventually TEE base images for other languages (NodeJs).

Reviewer 3: not sure to understand the second part
